mkosi/config.py

```diff
@@ -696,8 +700,17 @@ def from_json(cls, s: Union[str, dict[str, Any], SupportsRead[str], SupportsRead
             raise ValueError(f"{cls.__name__} can only be constructed from JSON from strings, dictionaries and files.")
 
         transformer = json_type_transformer(cls)
-        tj = {k: transformer(k, v) for k, v in j.items()}
-        return cls(**tj)
+        return {k: transformer(k, v) for k, v in j.items()}
+
+    @classmethod
+    def from_json(cls, s: Union[str, dict[str, Any], SupportsRead[str], SupportsRead[bytes]]) -> "MkosiArgs":
+        j = cls._load_json(s)
+        return cls(**j)
+
+    def merge_json(self, s: Union[str, dict[str, Any], SupportsRead[str], SupportsRead[bytes]]) -> "MkosiArgs":
```

Comment:
Do we need this method? What about having `from_json()` construct a default instance and then merging the json fields into that instance and returning that? If we have that I don't think we'd need a separate merge_json() method

Reply:
They serve two different purposes and I added docstrings to clarify that.

`from_json` will need a full JSON dump, since the constructor for `MKosiArgs`/`MkosiConfig` doesn't have defaults. The `empty` constructor will generate an empty `MkosiArgs`/`MkosiConfig` with default values. I think this is safer than adding defaults to the regular constructor since defaults will not be used implicitly. 

To accept partial JSON, which we will need later, the way would be `MkosiConfig.empty().merge_json(some_partial_json_string)`. It will probably make sense to roll `.empty().merge_json()`  both into `load_partial_json` method.

Reply:
Yes but when will we ever want to parse full JSON is what I'm wondering? I think we should just always allow partial JSON, to account for new properties that get added or maybe stuff that gets removed.

Reply:
I still think that parsing full JSON has a use for reproducible builds and when one explicitly doesn't want defaults applied implicitly. Right now this is a bit moot, because we still have to explore how this will be used. :)

Reply:
Added the `from_partial_json` method in this rebase.